Migrate test_complex_optimizer to OptimizerInfo

janeyx99 opened the pull request:
This PR does what it says and more.

1. We increase coverage by a LOT! Previously, complex was not tested for many many configs, including foreach + maximize at the same time. Or the fused impls. Or just random configs people forgot about.
2. I rearranged the maximize conditional and the _view_as_real to preserve list-ness. This is needed for _view_as_real to function properly, I did add a comment in the Files Changed. This new order also just...makes more aesthetic sense.
3. Note that LBFGS and SparseAdam are skipped--they don't support complex and now we know.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118161
* __->__ #118160

janeyx99 committed:
Migrate test_complex_optimizer to OptimizerInfo

[ghstack-poisoned]

janeyx99 commented on torch/optim/adadelta.py:
```diff
@@ -286,12 +286,12 @@ def _multi_tensor_adadelta(
 
     grouped_tensors = Optimizer._group_tensors_by_device_and_dtype([params, grads, square_avgs, acc_deltas])
     for ((device_params, device_grads, device_square_avgs, device_acc_deltas), _) in grouped_tensors.values():
-        if maximize:
```
Why are these these changes necessary?

It turns out we weren't testing complex with maximize before, and so we ran into failures when I turned on testing.
The crux is that foreach ops return tuples, which are not mutable, but the _`view_as_real` requires mutable lists. So when maximize came before, that turned `grouped_grads` into a tuple which was not good for `_view_as_real`.

janeyx99 commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -264,9 +264,7 @@ def get_error_inputs_for_all_optims(device, dtype):
         OptimizerInput(params=None, kwargs={}, desc="default"),
-        OptimizerInput(
-            params=None, kwargs={"lr": 0.01}, desc="non-default lr"
-        ),  # TODO: Move out to testing in param_group?
```
Remove old TODO, I've already landed param group tests.

janeyx99 commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -545,6 +544,21 @@ def optim_error_inputs_func_lbfgs(device, dtype):
 def optim_inputs_func_nadam(device=None):
     cuda_supported_configs = [
         OptimizerInput(params=None, kwargs={"capturable": True}, desc="capturable"),
+        OptimizerInput(
```
Add more configs for capturable for coverage--this _technically_ doesn't have to be in this PR...but if CI passes, then why not?

janeyx99 committed:
Update on "Migrate test_complex_optimizer to OptimizerInfo"

This PR does what it says and more.

1. We increase coverage by a LOT! Previously, complex was not tested for many many configs, including foreach + maximize at the same time. Or the fused impls. Or just random configs people forgot about.
2. I rearranged the maximize conditional and the _view_as_real to preserve list-ness. This is needed for _view_as_real to function properly, I did add a comment in the Files Changed. This new order also just...makes more aesthetic sense.
3. Note that LBFGS and SparseAdam are skipped--they don't support complex and now we know.




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/rmsprop.py:
```diff
@@ -348,14 +356,6 @@ def _multi_tensor_rmsprop(
 
         grouped_grads = list(grouped_grads)
```
I think this conversion to `list` can be removed maybe?

janeyx99 replied:
good catch!

mikaylagawarecki reviewed (approved):
sgtm

janeyx99 commented on test/test_optim.py:
```diff
@@ -128,6 +128,29 @@ def closure():
                 self.assertLess(closure().item(), initial_value)
 
 
+    @optims(optim_db, dtypes=[torch.complex64])
```
I am unsure, though any of the following:
- oversight
- coverage is sufficient
- we don't actually support complex128 nor chalf in certain ops so maybe they just don't work

A next step could be to try this out--it wouldn't be hard to enable the tests.

janeyx99 committed:
Update on "Migrate test_complex_optimizer to OptimizerInfo"

This PR does what it says and more.

1. We increase coverage by a LOT! Previously, complex was not tested for many many configs, including foreach + maximize at the same time. Or the fused impls. Or just random configs people forgot about.
2. I rearranged the maximize conditional and the _view_as_real to preserve list-ness. This is needed for _view_as_real to function properly, I did add a comment in the Files Changed. This new order also just...makes more aesthetic sense.
3. Note that LBFGS and SparseAdam are skipped--they don't support complex and now we know.




[ghstack-poisoned]

janeyx99 commented on test/test_optim.py:
```diff
@@ -128,6 +128,30 @@ def closure():
                 self.assertLess(closure().item(), initial_value)
 
 
+    @skipMPS
```
CI OS today too old to support complex (requires 14.1)
